pkg/cgroup/linux.go

```diff
@@ -0,0 +1,213 @@
+// after 3.x, please open a github issue https://github.com/minio/minio/issues
+// and discuss first if you wish to change this.
+const (
+	// Default value for any error or if cgroup is not set.
+	// this is the defautl if cgroups are not used/configured.
+	unlimitedMemoryLimit = 9223372036854771712
+
+	// Default string for looking for kernel memory param.
+	memoryLimitKernelParam = "memory.limit_in_bytes"
+
+	// Points to sys path memory path.
+	cgroupMemSysPath = "/sys/fs/cgroup/memory"
+
+	// Default docker prefix.
+	dockerPrefixName = "/docker/"
+
+	// Proc controller group path.
+	procCGroupPath = "/proc/%d/cgroup"
+)
+
+// CGEntries - represents all the entries in a process cgroup file
+// at /proc/<pid>/cgroup.
+type CGEntries map[string][]string
+
+// GetCGPathByName - returns the path matching the controller name.
+func (cg CGEntries) GetCGPathByName(cname string) (path string) {
+	return cg[cname][0]
+}
+
+// GetProcCGroup - reads and parses all the cgroup entries for a given process.
+func GetProcCGroup(pid int) (CGEntries, error) {
+	r, err := os.Open(fmt.Sprintf(procCGroupPath, pid))
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	return parseProcCGroup(r)
+}
+
+// ParseProcCGroup - cgroups are always in the following
+// format once enabled you need to know the pid of the
+// application you are looking for so that the the
+// following parsing logic only parses the file located
+// at /proc/<pid>/cgroup.
+//
+// CGROUP entries id, component and path are always in
+// the following format. ``ID:COMPONENT:PATH``
+//
+// Following code block parses this information and
+// returns a procCGroup which is a parsed list of all
+// the line by line entires from /proc/<pid>/cgroup.
+func parseProcCGroup(r io.Reader) (CGEntries, error) {
+	var proc = make(CGEntries)
+
+	// Start reading cgroup categories line by line
+	// and process them into procCGroup structure.
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		tokens := strings.SplitN(line, ":", 3)
+		if len(tokens) < 3 {
+			continue
+		}
+
+		_, name, path := tokens[0], tokens[1], tokens[2]
+		for _, token := range strings.Split(name, ",") {
+			name = strings.TrimPrefix(token, "name=")
+			proc[name] = []string{path}
+		}
+	}
+
+	// Return upon any error while reading the cgroup categories.
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return proc, nil
+}
+
+// Control group manager CLI defaults.
+const (
+	// Change this if at all in future control group manager
+	// changes their cli name.
+	cgmProgName = "cgm"
+
+	// Action param name used to fetch value for a particular
+	// controller.
+	cgmProgGetValue = "getvalue"
+)
+
+// Fetch value of the cgroup kernel param from the cgroup manager,
+// if cgroup manager is configured we should just rely on `cgm` cli
+// to fetch all the values for us.
+func getCGroupManagerKernValue(cname, path, kernParam string) (limit uint64, err error) {
```

Comment:
Does `cgm` gives other than `/sys/fs/cgroup/$MEMPATH/memory.limit_in_bytes`?
Is there a chance of missing `/sys/fs/cgroup/$MEMPATH/memory.limit_in_bytes` file?